File: algorithms/python/search/binary_search.py
def binary_search(arr, target):
    """
    Python implementation of Binary Search Algorithm.
    Requires the array to be sorted.
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    left, right = 0, len(arr) - 1
    
    while left <= right:
        mid = (left + right) // 2
        
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
            
    return -1

if __name__ == "__main__":
    sorted_arr = [2, 3, 4, 10, 40]
    target = 10
    
    result = binary_search(sorted_arr, target)
    
    if result != -1:
        print(f"Element is present at index {result}")
    else:
        print("Element is not present in array")

File: algorithms/kotlin/search/BinarySearch.kt
fun binarySearch(arr: IntArray, x: Int): Int {
    var l = 0
    var r = arr.size - 1
    while (l <= r) {
        val m = l + (r - l) / 2
        if (arr[m] == x) return m
        if (arr[m] < x) l = m + 1 else r = m - 1
    }
    return -1
}

fun main() {
    val arr = intArrayOf(2, 3, 4, 10, 40)
    val x = 10
    val result = binarySearch(arr, x)
    if (result == -1) println("Element is not present in array")
    else println("Element is present at index $result")
}
